src/AspNet.Security.OAuth.ExactOnline/ExactOnlineAuthenticationHandler.cs

```diff
@@ -0,0 +1,65 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using AuthenticationProperties = Microsoft.AspNetCore.Authentication.AuthenticationProperties;
+
+namespace AspNet.Security.OAuth.ExactOnline
+{
+    public class ExactOnlineAuthenticationHandler : OAuthHandler<ExactOnlineAuthenticationOptions>
+    {
+        public ExactOnlineAuthenticationHandler(
+            [NotNull] IOptionsMonitor<ExactOnlineAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+
+            using var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving the user profile.");
+            }
+
+            using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
+            var root = payload.RootElement.GetProperty("d").GetProperty("results")[0];
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, root);
+            context.RunClaimActions();
```

Comment:
What about something similar the below code from the Foursquare provider instead?

Then the `OAuthCreatingTicketContext` has access to the root response, rather than just the bit we've extracted from it, and we can also be a bit more defensive about checking the response, rather than assuming they array will have one item in it from the `[0]`? Then if something changed server side in the future and broke things it would be a bit easier to debug than throwing an `IndexOutOfRangeException` or similar.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/28f6c0d8114f3195c0897af60304bb12d37d9517/src/AspNet.Security.OAuth.Foursquare/FoursquareAuthenticationHandler.cs#L61-L73

Reply:
The entire API is an OData API, so this d/results/0 part is not specific for ExactOnline, but part of the OData standard. If they were to move away from OData, then I think many more things break. That's why I thought it was safe to keep it this way.

Reply:
In that case maybe something like this would be better. It makes it clearer that the code is expecting an array and also makes the entire JSON response availble to the `OAuthCreatingTicketContext`:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/28f6c0d8114f3195c0897af60304bb12d37d9517/src/AspNet.Security.OAuth.StackExchange/StackExchangeAuthenticationHandler.cs#L79-L80

So it would end up like:

```csharp
var user = payload.RootElement
    .GetProperty("d")
    .GetProperty("results")
    .EnumerateArray()
    .First();

var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
context.RunClaimActions(user);
```

Reply:
Ah, that is a good idea. I did not know the `RunClaimActions` accepted this as extra parameter.